sam_handler/handler.py

```diff
@@ -231,10 +243,33 @@ def _create_generation_destination(self):
         """Creates the destination_template_directory, if it doesn't exist."""
         self.destination_template_directory.mkdir(parents=True, exist_ok=True)
 
-    def _build(self, invoker: SamInvoker):
+    def _prepare_template(self) -> Path:
+        if self.sam_template_path.suffix not in self.supported_template_extensions:
+            raise UnsupportedTemplateFileTypeError(
+                f"Template has file extension {self.sam_template_path}. Only "
+                f"{self.supported_template_extensions} are supported."
+            )
+
+        if self.sam_template_path.suffix in self.standard_template_extensions:
+            return self.sam_template_path
+        elif self.sam_template_path.suffix in self.jinja_template_extensions:
+            return self._compile_jinja_template()
+
+    def _compile_jinja_template(self) -> Path:
+        self.logger.info("Compiling Jinja template...")
+        template_body = self.render_jinja_template(
+            str(self.sam_template_path),
+            {'sceptre_user_data': self.sceptre_user_data},
+            self.stack_group_config.get('j2_environment', {})
+        )
+        compiled_path = self.sam_template_path.parent / f'{self.sam_template_path.stem}.compiled'
```

Comment:
would be nice to document the compiled path so users know where to look for their generated templates.

The AWS SAM CLI puts generated artifacts in a hidden `.aws-sam` folder.  I'm wondering whether it would be better to put generated templates in a hidden folder instead?

Reply:
If you look above (https://github.com/Sceptre/sceptre-sam-handler/pull/3/files#diff-71099f4ab2e139b8ebc731a85e586a758ca70ef6d8f638eb330e8a7b8e115e79R197), we actually delete the compiled template right after we use it. I'm not sure we really need anything further in that regard, since the file literally exists for only about a minute.

Reply:
I didn't noticed that, thanks for pointing it out.  Hmm, I'm not sure that's desirable though.  What if someone wants to inspect the generated template before deploying it?  Part of my workflow when implementing jinja templates is to use the `sceptre generate` command to check that the generated template looks correct before doing the actual deployment.  Does this template handler work with `sceptre generate`?

Reply:
Yeah, running `sceptre generate` would work. Don't forget, the "compiled" file is really an intermediate state artifact.

If you used `sceptre generate` on a jinja SAM template would mean:
1. The jinja would be rendered to a template file (the one I clean up)
2. The rendered template would be passed to `sam build`, which would build the lambdas and such and create a build artifact in `.aws-sam`
3. The handler will run `sam package` which will upload compiled artifacts and update the template at `.aws-sam` to reference those S3 paths of the artifacts. We output the packaged templates  to a tempfile anyway
4. The handler will return the contents of that tempfile.

In other words, there's really 3 stages of successive transformations:
* Stage 1 (the compiled jinja) is deleted after the process so as not to clutter the directory
* Stage 2 (initial build) gets moved to `.aws-sam/` and will be overwritten each time a build happens.
* Stage 3 (sam packaging) gets output to a temp file for the purposes of reading in its contents

Thus, if you wanted to see the generated template, you really want the output of Stage 3 (which is what `sceptre generate` will output. I see no real need to inspect the stage 1 output there. I suppose I could add a handler argument like `save_compiled_jinja` or something to not delete it... but I don't really see much use for that apart from potentially debugging a bad jinja template that SAM is rejecting. Would you like me to add that flag as an argument?

Reply:
nah, as long as users can check generated jinja output with sceptre generate then i think it's fine as-is.